### ISSUE TITLE
OpenAudible downloads fail with "Connection reset" error.

### ISSUE THREAD

**Issue by Blueliner59** [openaudible/openaudible#676]
Hello,

I have been trying for two days to download/backup my Audible library.  In that time I have successfully downloaded 24 out of 275 books.  A few books fail downloading with an error message stating "Read timed out".  Others fail with a message of "Download error: no library record" - there are perhaps a dozen of those and they appear to be the "free" books from Audible - so perhaps they have a different status within the Audible database???  

However, most of the failed downloads give an error message stating "Connection reset".  I have a very slow satellite internet connection (with a few disconnects/reconnects per day) and I assume this may be related to that.  Perhaps a timeout issue(?).  I have tried downloading with and without the VPN but get the same result.  I have also tried fully re-syncing my library and turning off all other programs running on the computer in question (an Intel NUC running Arch Linux).  OpenAudible appears to be running fine - I have seen no errors in the computer logs, all functions work (except for the download failures) and the gui response is good.  The downloads are going to a network NAS and both the computer and the NAS are connected to the same router via Ethernet cable.  OpenAudible diagnostics reports all directories exist and are available.  The NAS is used almost exclusively for multi-media and performs well with all applications I use (and currently has about 2 TB of free space).  Downloads from other programs, including those to the NAS, are working well and never fail. 

Do you have any suggestions on how this can be corrected?

Thank you

**Comment by openaudible** [comment]
No library record is usually because it is a podcast or other "free" or included book. Those are not something we support backing up.

But the read timeout is unfortunate and may be related to your slow satellite connection. Try again and see if it works on a second or 3rd time. You could also grab the download URL from the log window and try to get the AAX file from a web browser.. and it that works, drag and drop the file onto OpenAudible.

**Comment by Blueliner59** [comment]
> No library record is usually because it is a podcast or other "free" or included book. Those are not something we support backing up.
> 
> But the read timeout is unfortunate and may be related to your slow satellite connection. Try again and see if it works on a second or 3rd time. You could also grab the download URL from the log window and try to get the AAX file from a web browser.. and it that works, drag and drop the file onto OpenAudible.

Hello,

Thank you for the suggestion to manually download and import.  It woks but is very slow and inconvenient.  Would it be possible to show the Audible download link in the book detail window or perhaps providing a "download  from default browser" option/link in the Audible book list?  A multiple retry setting for downloads may help too.

Anyway, it may take a while but I am getting the books downloaded, and the import/conversion is working flawlessly.

Thank you!

**Comment by openaudible** [comment]
We set the read/connect timeout to 300 seconds. That seems like a long time to go without getting a single byte of data. I'm not sure if making it a longer timeout would accomplish much, but I'd be willing to try to set it to "no timeout". Maybe you could send an error report of any of your "timeout" or connection reset messages.  I assume manually retrying the download should help-- maybe run it at night when there are fewer people connected to your ISP?

**Comment by Blueliner59** [comment]
Hello,

I agree that 300 seconds should be plenty long.  Looking more closely, the most common error is "Connection Reset" and that can happen at any time during a download.  I started downloads just now (~11pm) and the first one failed with "Connection Reset" within 20 seconds.  I see several Java socket / SSL errors.  Downloading manually from Firefox worked most of the time but there has been  instances where I had to restart a download once or twice.

Not sure whats up but I have included a copy of the console log.  I am not that knowledgeable with this sort of thing, so any suggestions would be appreciated.

Thank you

[Console.txt](https://github.com/openaudible/openaudible/files/6920990/Console.txt)

**Comment by openaudible** [comment]
A few things to try: 

Only download 1 book at a time. Go to OA preferences dialog, click "Show Settings", navigate to OpenAudible directory and open settings.json with a text editor. Change concurrentDownloads:3 to concurrentDownloads:1 and save and restart OA.  That will reduce the amount of data going over the network at one time.

I posted a new beta for you that should retry downloads when they fail (up to 4 times.) I also added more logging and updated the httpclient library to the latest version.  Give it a try and send another copy of your logs if you have errors to report. 

https://openaudible.org/beta

**Comment by Blueliner59** [comment]
Hello,

I haven't been home much but I did install the beta and gave it a try.  Downloads failed at the same rate.  I could not notice a difference in how the beta responded to failed downloads but I didn't really know what to look for.  I checked the network settings and configurations on the Arch device that was failing and my Manjaro Linux laptop....they were the same.  I then tried OpenAudible beta on the laptop and connected to a different network - everything worked perfectly!

I made a few network/tcp settings changes on the Arch device.  It will now successfully download if it is not connected to the VPN.  It looks like I will need to do some digging into my internet/network/VPN.  (Any ideas on that?)

Thank you for taking the time to develop the beta for me to try.  If the console logs would still be helpful, I can send them when I get home again in a few days.  

Thank you
